### PR TITLE
Upgrade etcd to 3.5.0

### PR DESCRIPTION
**Pull request by Aceralon**
## Description

Upgrade etcd to 3.5.0. There are many projects import libcalico-go, if those projects need to upgrade to etcd 3.5.0, libcalico-go must use etcd 3.5.0 as well. 

## Todos
- [ ] Tests
- [ ] Documentation
- [x] Release note

## Release Note

```release-note
Update etcd client version to v3.5.0
```

### CODE REVIEW

**Comment by caseydavenport** [comment]
> There are many projects import libcalico-go,

I'd recommend against importing libcalico-go. It's very likely that you want github.com/projectcalico/api instead.

That said, I don't see anything wrong with bumping the etcd version.

File: go.mod [hunk 1]
```diff
@@ -39,4 +39,7 @@ require (
 	sigs.k8s.io/kind v0.11.1
 )
 
-replace github.com/sirupsen/logrus => github.com/projectcalico/logrus v1.0.4-calico
+replace (
+	github.com/sirupsen/logrus => github.com/projectcalico/logrus v1.0.4-calico
+	google.golang.org/grpc => google.golang.org/grpc v1.29.1
```
Comment (caseydavenport) [inline go.mod:44]: Why do we need to replace the grpc library version?

Comment (Aceralon) [reply go.mod:44]: The latest version of `google.golang.org/grpc` that has `google.golang.org/grpc/naming` is `v1.29.1`.
The old version of etcd like v3.3.25 uses 'naming'. 
```
go: finding module for package google.golang.org/grpc/naming
github.com/projectcalico/libcalico-go/lib/backend/etcd imports
        go.etcd.io/etcd/client tested by
        go.etcd.io/etcd/client.test imports
        github.com/coreos/etcd/integration imports
        github.com/coreos/etcd/proxy/grpcproxy imports
        google.golang.org/grpc/naming: module google.golang.org/grpc@latest found (v1.40.0), but does not contain package google.golang.org/grpc/naming
```

**Comment by caseydavenport** [comment]
@Aceralon looks like this PR needs a rebase

**Comment by Aceralon** [comment]
> @Aceralon looks like this PR needs a rebase

Rebased.

**Comment by caseydavenport** [comment]
Looks like static-checks are failing:

> lib/validator/v3/validator_test.go:29:5: could not import github.com/projectcalico/libcalico-go/lib/validator/v3 (lib/validator/v3/validator.go:34:9: could not import github.com/projectcalico/libcalico-go/lib/apis/v3 (lib/apis/v3/openapi_generated.go:110:17: cannot use ref("github.com/projectcalico/libcalico-go/lib/apis/v1/unversioned.TypeMetadata") (value of type spec.Ref) as spec.Ref value in struct literal)) (typecheck)02:07
	v3 "github.com/projectcalico/libcalico-go/lib/validator/v3"

I suspect this is because the PR reverts the version of `github.com/projectcalico/api` to an earlier commit, which isn't compatible with the latest.

`make PIN_BRANCH=master update-pins` should correct it, or you can manually adjust the api pin to what it was before.

**Comment by Aceralon** [comment]
> Looks like static-checks are failing:
> 
> > lib/validator/v3/validator_test.go:29:5: could not import github.com/projectcalico/libcalico-go/lib/validator/v3 (lib/validator/v3/validator.go:34:9: could not import github.com/projectcalico/libcalico-go/lib/apis/v3 (lib/apis/v3/openapi_generated.go:110:17: cannot use ref("github.com/projectcalico/libcalico-go/lib/apis/v1/unversioned.TypeMetadata") (value of type spec.Ref) as spec.Ref value in struct literal)) (typecheck)02:07
> > v3 "github.com/projectcalico/libcalico-go/lib/validator/v3"
> 
> I suspect this is because the PR reverts the version of `github.com/projectcalico/api` to an earlier commit, which isn't compatible with the latest.
> 
> `make PIN_BRANCH=master update-pins` should correct it, or you can manually adjust the api pin to what it was before.

Done.

File: go.mod [hunk 2]
```diff
@@ -34,9 +38,14 @@ require (
 	k8s.io/api v0.21.0-rc.0
 	k8s.io/apimachinery v0.21.0-rc.0
 	k8s.io/client-go v0.21.0-rc.0
-	k8s.io/code-generator v0.21.0-rc.0
-	k8s.io/kube-openapi v0.0.0-20210305001622-591a79e4bda7
+	k8s.io/code-generator v0.22.1
```
Comment (caseydavenport) [inline go.mod:41]: I think these changes are still wrong and need to be reverted

Comment (Aceralon) [reply go.mod]: Changed back to v0.21.0-rc.0

**Comment by caseydavenport** [comment]
Had to revert this for now because it broke downstream builds: https://github.com/projectcalico/libcalico-go/pull/1506

e.g., https://tigera.semaphoreci.com/jobs/aff600b6-528d-488b-9ae8-a716991e9a27

**Comment by Aceralon** [comment]
> Had to revert this for now because it broke downstream builds: #1506
> 
> e.g., https://tigera.semaphoreci.com/jobs/aff600b6-528d-488b-9ae8-a716991e9a27

I think maybe the problem is that we should manually replace grpc by `google.golang.org/grpc => google.golang.org/grpc v1.29.1` in https://github.com/projectcalico/typha/blob/master/go.mod

For the error in https://tigera.semaphoreci.com/jobs/aff600b6-528d-488b-9ae8-a716991e9a27
```
github.com/projectcalico/typha/pkg/daemon imports00:59
	github.com/projectcalico/libcalico-go/lib/upgrade/migrator/clients imports00:59
	github.com/projectcalico/libcalico-go/lib/upgrade/migrator/clients/v1/etcdv2 imports00:59
	go.etcd.io/etcd/client tested by00:59
	go.etcd.io/etcd/client.test imports00:59
	github.com/coreos/etcd/integration imports00:59
	github.com/coreos/etcd/proxy/grpcproxy imports00:59
	google.golang.org/grpc/naming: module google.golang.org/grpc@latest found (v1.40.0), but does not contain package google.golang.org/grpc/naming
```

**Pull request merged**